### PR TITLE
feat(Animated): Adds native animation drivers for ReactWindows

### LINKED ISSUE #566: Add Native Animation Features
I.e., this stuff:
https://github.com/facebook/react-native/tree/master/ReactAndroid/src/main/java/com/facebook/react/animated

### PR DESCRIPTION
**Pull request by rozele**
Android has a native implementation of the Animated graph.  This initial commit adds the same for Windows.

Fixes #566

### CODE REVIEW

File: ReactWindows/ReactNative/Animated/AdditionAnimatedNode.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+                {
+                    throw new InvalidOperationException(
+                        "Illegal node ID set as an input for Animated.Add node.");
+                }
+
+                Value += valueNode.Value;
+            }
+        }
+    }
+}
```
Comment (rozele) [inline ReactWindows/ReactNative/Animated/AdditionAnimatedNode.cs]: Double check if we generally add new lines to the ends of files...

File: ReactWindows/ReactNative/Animated/AdditionAnimatedNode.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    class AdditionAnimatedNode : ValueAnimatedNode
+    {
+        private readonly NativeAnimatedNodesManager _manager;
+        private readonly int[] _inputNodes;
+
+        public AdditionAnimatedNode(int tag, JObject config, NativeAnimatedNodesManager manager)
+            : base(tag)
+        {
+            _manager = manager;
+            _inputNodes = config.GetValue("input").ToObject<int[]>();
```
Comment (mpodwysocki) [inline ReactWindows/ReactNative/Animated/AdditionAnimatedNode.cs]: Is there a null check we need to do here?

File: ReactWindows/ReactNative/Animated/AnimatedNode.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+        public IReadOnlyList<AnimatedNode> Children
+        {
+            get
+            {
+                return _children;
+            }
+        }
+
+        public long BFSColor
```
Comment (mpodwysocki) [inline ReactWindows/ReactNative/Animated/AnimatedNode.cs]: Does this follow the .NET naming standards ala `BfsColor`?

File: ReactWindows/ReactNative/Animated/FrameBasedAnimationDriver.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        private readonly double[] _frames;
+        private readonly double _toValue;
+
+        private long _startFrameTimeTicks = -1;
+        private double _fromValue;
+
+        public FrameBasedAnimationDriver(int id, ValueAnimatedNode animatedValue, ICallback endCallback, JObject config)
+            : base(id, animatedValue, endCallback)
+        {
+            _frames = config.GetValue("frames").ToObject<double[]>();
```
Comment (mpodwysocki) [inline ReactWindows/ReactNative/Animated/FrameBasedAnimationDriver.cs]: any time dealing with JSON, do we need null checks?

File: ReactWindows/ReactNative/Animated/InterpolationAnimatedNode.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        private readonly double[] _inputRange;
+        private readonly double[] _outputRange;
+
+        private ValueAnimatedNode _parent;
+
+        public InterpolationAnimatedNode(int tag, JObject config)
+            : base(tag)
+        {
+            _inputRange = config.GetValue("inputRange").ToObject<double[]>();
```
Comment (mpodwysocki) [inline ReactWindows/ReactNative/Animated/InterpolationAnimatedNode.cs]: Same comment as above

File: ReactWindows/ReactNative/Animated/SpringAnimationDriver.cs [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+            double bVelocity, bAcceleration;
+            double cVelocity, cAcceleration;
+            double dVelocity, dAcceleration;
+
+            double dxdt, dvdt;
+
+            // iterate over the true time
+            while (_timeAccumulator >= SolverTimestepSec)
+            {
+                /* begin debug
```
Comment (mpodwysocki) [inline ReactWindows/ReactNative/Animated/SpringAnimationDriver.cs]: Feel free to delete

File: ReactWindows/ReactNative/Animated/SpringAnimationDriver.cs [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+                {
+                    _endValue = _currentState.Position;
+                    _startValue = _endValue;
+                }
+
+                _currentState.Velocity = 0;
+            }
+        }
+
+        class PhysicsState
```
Comment (mpodwysocki) [inline ReactWindows/ReactNative/Animated/SpringAnimationDriver.cs]: Maybe use a struct here?

Comment (rozele) [inline ReactWindows/ReactNative/Animated/SpringAnimationDriver.cs]: Yeah, I suppose non-readonly structs would work.

---
In reply to: [74450042](https://github.com/ReactWindows/react-native-windows/pull/583#discussion_r74450042) [](ancestors = 74450042)

File: ReactWindows/ReactNative/Animated/AdditionAnimatedNode.cs (hunk 2 above)
Comment (rozele) [inline ReactWindows/ReactNative/Animated/AdditionAnimatedNode.cs]: JSON validation should be done in JavaScript.

---
In reply to: [74442452](https://github.com/ReactWindows/react-native-windows/pull/583#discussion_r74442452) [](ancestors = 74442452)

File: ReactWindows/ReactNative/Animated/AnimatedNode.cs (hunk 3 above)
Comment (rozele) [inline ReactWindows/ReactNative/Animated/AnimatedNode.cs]: We should get a stylecop GitHub plugin that makes snarky comments about .NET naming standards :-D

---
In reply to: [74442832](https://github.com/ReactWindows/react-native-windows/pull/583#discussion_r74442832) [](ancestors = 74442832)

File: ReactWindows/ReactNative/Animated/FrameBasedAnimationDriver.cs (hunk 4 above)
Comment (rozele) [inline ReactWindows/ReactNative/Animated/FrameBasedAnimationDriver.cs]: Generally not worried about it, It's not user-generated JSON, the data validation should occur JS side in most cases.

---
In reply to: [74443274](https://github.com/ReactWindows/react-native-windows/pull/583#discussion_r74443274) [](ancestors = 74443274)

**Pull request merged**